packages/ERTP/src/amountMath.js

```diff
@@ -54,138 +54,188 @@ export { MathKind };
- * makeLocalAmountMath(issuer).
- *
- * AmountMath exports MathKind, which contains constants for the kinds:
- * NAT, SET, and STRING_SET.
- *
  * Each issuer of digital assets has an associated brand in a one-to-one
  * mapping. In untrusted contexts, such as in analyzing payments and
  * amounts, we can get the brand and find the issuer which matches the
  * brand. The issuer and the brand mutually validate each other.
- *
- * @param {Brand} brand
- * @param {AmountMathKind} amountMathKind
- * @returns {AmountMath}
  */
-function makeAmountMath(brand, amountMathKind) {
+
+/** @type {{ nat: NatMathHelpers, set: SetMathHelpers, strSet: SetMathHelpers }} */
+const helpers = {
+  nat: natMathHelpers,
+  set: setMathHelpers,
+  strSet: setMathHelpers,
+};
+
+/**
+ * @type {(value: NatValue | SetValue) => SetMathHelpers | NatMathHelpers }
+ */
+const getHelpersFromValue = value => {
+  if (Array.isArray(value)) {
+    return setMathHelpers;
+  }
+  assert(
+    typeof Nat(value) === 'bigint',
+    X`value ${value} must be a bigint or an array`,
+  );
+  return natMathHelpers;
+};
+
+/** @type {(amount: Amount ) => NatMathHelpers | SetMathHelpers} */
+const getHelpersFromAmount = amount => {
+  // @ts-ignore
+  return getHelpersFromValue(amount.value);
+};
+
+/** @type {(leftAmount: Amount, rightAmount: Amount ) =>
+ * NatMathHelpers | SetMathHelpers } */
+const getHelpers = (leftAmount, rightAmount) => {
+  const leftHelpers = getHelpersFromAmount(leftAmount);
+  const rightHelpers = getHelpersFromAmount(rightAmount);
+  assert.equal(leftHelpers, rightHelpers);
+  return leftHelpers;
+};
+
+/** @type {(amount: Amount, brand?: Brand) => void} */
+const optionalBrandCheck = (amount, brand) => {
+  if (brand !== undefined) {
+    mustBeComparable(brand);
+    assert.equal(
+      amount.brand,
+      brand,
+      X`amount's brand ${amount.brand} did not match expected brand ${brand}`,
+    );
+  }
+};
+
+/** @type {(value: Value, brand: Brand) => Amount} */
+const noCoerceMake = (value, brand) => {
+  const amount = harden({ brand, value });
+  return amount;
+};
+
+/** @type {(value: Value) => void} */
+const assertLooksLikeValue = value => {
+  assert(
+    Array.isArray(value) || isNat(value),
+    X`value ${value} must be a Nat or an array`,
+  );
+};
+
+const brandMethods = ['isMyIssuer', 'getAllegedName', 'getDisplayInfo'];
+
+const checkBrand = (brand, msg) => {
   mustBeComparable(brand);
-  assert.typeof(amountMathKind, 'string');
-
-  const mathHelpers = {
-    nat: natMathHelpers,
-    strSet: strSetMathHelpers,
-    set: setMathHelpers,
-  };
-  const helpers = mathHelpers[amountMathKind];
+  assert(typeof brand === 'object', msg);
+  const ownKeys = Reflect.ownKeys(brand);
+  const inBrandMethods = key => brandMethods.includes(key);
   assert(
-    helpers !== undefined,
-    X`unrecognized amountMathKind: ${amountMathKind}`,
+    passStyleOf(brand) === REMOTE_STYLE && ownKeys.every(inBrandMethods),
```

Comment:
Interesting. I would not have thought to check that.

Reply:
> passStyleOf(brand) === REMOTE_STYLE &&

Having fixed the code on line 128, this part of the check on line 132 is redundant and should be removed.

Reply:
op, good eye. I will remove

Reply:
See https://github.com/Agoric/agoric-sdk/pull/2597